pkg/loadbalancer/experimental/healthserver.go

```diff
@@ -0,0 +1,215 @@
+
+	lb "github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/rate"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// healthServerCell implements the support for the HealthCheckNodePort field.
+// For each service that has HealthCheckNodePort set to a non-zero value it
+// starts an HTTP server on that port and responds with the number of active
+// backends. This allows external load-balancers to avoid directing traffic
+// to services that have no available backends.
+var healthServerCell = cell.Module(
+	"healthserver",
+	"Serves NodePort health status to external load-balancers",
+	cell.Invoke(registerHealthServer),
+)
+
+type healthServerParams struct {
+	cell.In
+
+	Jobs      job.Group
+	Log       *slog.Logger
+	DB        *statedb.DB
+	ExtConfig ExternalConfig
+	Services  statedb.Table[*Service]
+	Backends  statedb.Table[*Backend]
+}
+
+// healthServer manages HTTP health check ports. For each added service,
+// it opens a HTTP server on the specified HealthCheckNodePort and either
+// responds with 200 OK if there are local endpoints for the service, or with
+// 503 Service Unavailable if the service does not have any local endpoints.
+type healthServer struct {
+	params        healthServerParams
+	serverByPort  map[uint16]*httpHealthServer
+	portByService map[lb.ServiceName]uint16
+}
+
+func registerHealthServer(params healthServerParams) {
+	if !params.ExtConfig.EnableHealthCheckNodePort {
+		return
+	}
+
+	s := &healthServer{
+		params:        params,
+		serverByPort:  map[uint16]*httpHealthServer{},
+		portByService: map[lb.ServiceName]uint16{},
+	}
+	params.Jobs.Add(job.OneShot("control-loop", s.controlLoop))
+}
+
+func (s *healthServer) controlLoop(ctx context.Context, health cell.Health) error {
+	// Watch services for changes to add and remove the listeners.
+	wtxn := s.params.DB.WriteTxn(s.params.Services)
+	serviceChanges, err := s.params.Services.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	// Limit the rate at which the change batches are processed.
+	limiter := rate.NewLimiter(100*time.Millisecond, 1)
+
+	defer s.cleanupListeners(ctx)
+
+	for {
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+
+		changes, watch := serviceChanges.Next(s.params.DB.ReadTxn())
+
+		for change := range changes {
+			svc := change.Object
+			name := svc.Name
+			port := svc.HealthCheckNodePort
+
+			if change.Deleted {
+				if port, ok := s.portByService[name]; ok {
+					s.removeListener(ctx, port)
+					delete(s.portByService, name)
+				}
+			} else {
+				oldPort, found := s.portByService[name]
+				if found && oldPort != port {
+					// HealthCheckNodePort has changed, we treat this as a DeleteService
+					// followed by an Insert.
+					s.removeListener(ctx, oldPort)
+					delete(s.portByService, name)
+					found = false
+				}
+				if port == 0 || found {
+					continue
+				}
+				s.serverByPort[port] = s.addListener(name, port)
+				s.portByService[name] = port
+			}
```

Comment:
I got hung up on this a bit, that `found = false` part is difficult :slightly_smiling_face:.
I broke it down for myself into what's below to confirm the logic, maybe you want to pick it up?

```suggestion
				oldPort, found := s.portByService[name]
			       // HealthCheckNodePort has changed, we treat this as a DeleteService
			       // followed by an Insert.

				if found && oldPort == port {
				        // no change needed
				        continue
				}

                               // delete old port
			       if (found) {
                                        s.removeListener(ctx, oldPort)
					delete(s.portByService, name)
				}

                               // add new port
				if (port) {
				        s.serverByPort[port] = s.addListener(name, port)
				        s.portByService[name] = port
			       }
```

Reply:
Thanks, reworked it!